src/components/Web3Modal.tsx

```diff
@@ -1,127 +1,167 @@
-import InitialExplorer from '../views/InitialExplorer';
-import ViewAllExplorer from '../views/ViewAllExplorer';
-import QRCodeView from '../views/QRCodeView';
-
-const INITIAL_ROUTE = 'INIT_WALLETS';
+import { createUniversalProvider, createSession } from '../utils/ProviderUtil';
+import { ModalCtrl } from '../controllers/ModalCtrl';
+import { Web3ModalRouter } from './Web3ModalRouter';
+import { ExplorerCtrl } from '../controllers/ExplorerCtrl';
+import { ConfigCtrl } from '../controllers/ConfigCtrl';
+import { OptionsCtrl } from '../controllers/OptionsCtrl';
+import { ClientCtrl } from '../controllers/ClientCtrl';
 
 interface Web3ModalProps {
   projectId: string;
-  isVisible: boolean;
-  onClose: () => void;
-  currentWCURI?: string;
+  relayUrl?: string;
+  onCopyClipboard?: (value: string) => void;
 }
 
-export function Web3Modal({ projectId, isVisible, onClose }: Web3ModalProps) {
-  const [isWalletListLoading, setWalletListLoading] = useState(true);
-  const [initialWallets, setInitialWallets] = useState<Listing[]>([]);
-  const [allWallets, setAllWallets] = useState<Listing[]>([]);
-
+export function Web3Modal({
+  projectId,
+  relayUrl,
+  onCopyClipboard,
+}: Web3ModalProps) {
+  const modalState = useSnapshot(ModalCtrl.state);
   const isDarkMode = useColorScheme() === 'dark';
 
-  const [viewStack, setViewStack] = useState<Routes[]>([INITIAL_ROUTE]);
+  const onSessionCreated = useCallback(async () => {
+    OptionsCtrl.getAccount();
+    ModalCtrl.close();
+  }, []);
+
+  const onSessionError = useCallback(async () => {
+    ModalCtrl.close();
+    Alert.alert('Error', 'Error with session');
+  }, []);
+
+  const onSessionDelete = useCallback(async ({ topic }: { topic: string }) => {
+    const session = ClientCtrl.session();
+    if (topic === session?.topic) {
+      OptionsCtrl.resetAccount();
+      ClientCtrl.clearSession();
+    }
+  }, []);
 
-  const fetchWallets = useCallback(() => {
-    ExplorerUtil.fetchWallets(projectId).then((wallets) => {
-      setWalletListLoading(false);
-      if (wallets) {
-        setInitialWallets(wallets.listings.slice(0, 7));
-        setAllWallets(wallets.listings);
+  const onConnect = useCallback(async () => {
+    const provider = ClientCtrl.provider();
+    try {
+      const session = await createSession(provider);
+      if (session) {
+        ClientCtrl.setSession(session);
+        onSessionCreated();
       }
-    });
-  }, [projectId]);
+    } catch (error) {
+      onSessionError();
+    }
+  }, [onSessionCreated, onSessionError]);
 
-  const onNavigate = useCallback(
-    (route: Routes) => {
-      setViewStack(viewStack.concat([route]));
-    },
-    [viewStack]
-  );
+  const subscribeToEvents = useCallback(async () => {
+    const provider = ClientCtrl.provider();
+    if (provider) {
+      provider.on('display_uri', (uri: string) => {
+        OptionsCtrl.setSessionUri(uri);
+      });
+
+      // Subscribe to session ping
+      provider.on(
```

Comment:
Should these be unsubscribed from somewhere i.e. instead of useCallback, perhaps put them in useEffect and remove when unmounting?

Reply:
you are right, i didn't know there was a function to remove those listeners. Should we add something in the [docs](https://docs.walletconnect.com/2.0/javascript/providers/universal#events)?